Sources/Plasma/PubUtilLib/plPipeline/plTextFont.cpp

```diff
@@ -104,183 +98,137 @@ uint16_t  *plTextFont::IInitFontTexture()
+    
     hFont = CreateFont( nHeight, 0, 0, 0, FW_NORMAL, FALSE, FALSE, FALSE, DEFAULT_CHARSET, OUT_DEFAULT_PRECIS,
                         CLIP_DEFAULT_PRECIS, ANTIALIASED_QUALITY, VARIABLE_PITCH, fFace );
     hsAssert(hFont != nullptr, "Cannot create Windows font");
 
-    SelectObject( hDC, hBitmap );
-    SelectObject( hDC, hFont );
-
-    // Set text colors
-    SetTextColor( hDC, RGB( 255, 255, 255 ) );
-    SetBkColor( hDC, 0 );
-    SetTextAlign( hDC, TA_TOP );
-
-    // Loop through characters, drawing them one at a time
-    RECT    r;
-    r.left = r.top = 0;
-    r.right = r.bottom = 10;
-    FillRect( hDC, &r, (HBRUSH)GetStockObject( GRAY_BRUSH ) );
-
-    // (Make first character a black dot, for filling rectangles)
-    SetPixel( hDC, 0, 0, RGB( 255, 255, 255 ) );
-    for( c = 32, x = 1, y = 0; c < 127; c++ )
-    {
-        myChar[ 0 ] = c;
-        GetTextExtentPoint32( hDC, myChar, 1, &size );
-
-        if( (uint32_t)( x + size.cx + 1 ) > fTextureWidth )
-        {
-            x = 0;
-            y += size.cy + 1;
-        }
-
-        ExtTextOut(hDC, x, y, ETO_OPAQUE, nullptr, myChar, 1, nullptr);
-
-        fCharInfo[ c ].fW = (uint16_t)size.cx;
-        fCharInfo[ c ].fH = (uint16_t)size.cy;
-        fCharInfo[ c ].fUVs[ 0 ].fX = (float)x / (float)fTextureWidth;
-        fCharInfo[ c ].fUVs[ 0 ].fY = (float)y / (float)fTextureHeight;
-        fCharInfo[ c ].fUVs[ 1 ].fX = (float)( x + size.cx ) / (float)fTextureWidth;
-        fCharInfo[ c ].fUVs[ 1 ].fY = (float)( y + size.cy ) / (float)fTextureHeight;
-        fCharInfo[ c ].fUVs[ 0 ].fZ = fCharInfo[ c ].fUVs[ 1 ].fZ = 0;
-
-        x += size.cx + 1;
-    }
-    fCharInfo[ 32 ].fUVs[ 1 ].fX = fCharInfo[ 32 ].fUVs[ 0 ].fX;
-
-    // Special case the tab key
-    fCharInfo[ '\t' ].fUVs[ 1 ].fX = fCharInfo[ '\t' ].fUVs[ 0 ].fX = fCharInfo[ 32 ].fUVs[ 0 ].fX;
-    fCharInfo[ '\t' ].fUVs[ 1 ].fY = fCharInfo[ '\t' ].fUVs[ 0 ].fY = 0;
-    fCharInfo[ '\t' ].fUVs[ 0 ].fZ = fCharInfo[ '\t' ].fUVs[ 1 ].fZ = 0;
-    fCharInfo[ '\t' ].fW = fCharInfo[ 32 ].fW * 4;
-    fCharInfo[ '\t' ].fH = fCharInfo[ 32 ].fH;
-
-    /// Now create the data block
-    uint16_t  *data = new uint16_t[ fTextureWidth * fTextureHeight ];
-    tBits = data;
-    for( y = 0; y < fTextureHeight; y++ )
-    {
-        for( x = 0; x < fTextureWidth; x++ )
-        {
-            bAlpha = (BYTE)( ( bitmapBits[ fTextureWidth * y + x ] & 0xff ) >> 4 );
-
-            if( bitmapBits[ fTextureWidth * y + x ] )
-                *tBits = 0xffff;
-            else
-                *tBits = 0;
-
-            tBits++;
-        }
-    }
+    SelectObject(hDC, hFont);
+    
+    DWORD fontDataSize = GetFontData( hDC, 0, 0, 0, 0 );
+    void* fontData = std::malloc( fontDataSize );
+    GetFontData(hDC, 0, 0, fontData, fontDataSize);
+    ftError = FT_New_Memory_Face(library, (FT_Byte *) fontData, fontDataSize, 0, &face);
+    FT_UInt freeTypeResolution = GetDeviceCaps(hDC, LOGPIXELSY);
```

Comment:
I don't know a better way to load the font on Windows other than actually loading it as a Win32 font and pulling the TTF data. Didn't see anything better on their internet either for Win32.

However: There may be better options in things like DirectWrite. I had avoided other options because of the XP-as-baseline policy.

Reply:
Windows 7 APIs are fine to use

Reply:
After some more looking, I'm inclined to leave these GDI calls in place at least for the short term. Windows has the complication that TTFs are not guaranteed to be files, so they may not have a file path. I don't think that applies to Courier as shipped, but the copy-font-as-data path may not be as wrong as it feels on Windows. But I'm guessing this might be revisited in the future.